rfcs/framework-0063-notification-system.md

```diff
@@ -0,0 +1,422 @@
+        mail_data["html_body"] = template_html.render(**ctx)
+        mail_data["plain_body"] = template_txt.render(**ctx)
+        resp = send_email(mail_data)
+        return resp  # TODO: what would a "delivery" result be
+```
+
+#### Templating
+
+A generic Jinja template loader should make life easier for backends to integrate templates and take into account locale:
+
+```python
+class JinjaTemplateLoaderMixin:
+    """Used only in NotificationBackend classes."""
+
+    template_folder = 'notifications'
+
+    def _load_template(path):
+        try:
+            template = current_app.jinja_env.get_template(path)
+        except TemplateNotFoundError:
+            template = None
+        return template
+
+    def get_template(self, recipient: Recipient):
+        # e.g /notifications/email/comment_edit.html
+        base_template = self._load_template(f"{template_folder}/{recipient.template}")
+        specific_template = self._load_template(f"{template_folder}/{self.id}/{recipient.template}")
+
+        template = specific_template or base_template
+        if not template:
+            raise TemplateNotFoundError()
+
+        # Take locale into account
+        locale = recipient.data.get("locale")
+        if locale != "en":
+            template += f".{locale}"
+
+        return template
+```
+
+### Building notifications
+
+The API described above, accepts a `Notification` object as its main payload. Generating this object is something that is left to the client, which in our case is usually a service method. We want to provide a set of utilities that will streamline this process and make it configurable.
+
+Another consideration is that different parts of the platform contribute different business rules for:
+
+- types of notifications and their templates
+- the set of recipients for a notification
+- individual notification preferences for each recipient
+- what notification backends (delivery methods) should be used for a recipient
+
+To encapsulate all this functionality we're introducing a set of classes that can be composed to build individual notification builders for each type of notification.
+
+```python
+class NotificationBuilder:
+
+    payload: list[PayloadGenerator]
+    recipients: list[RecipientGenerator]
+    backends: list[BackendGenerator]
+
+    def build(context: dict) -> Notification:
+        ...
+```
+
+Notification builder implementation for community record inclusion request:
+
+```python
+# invenio_requests/notifications.py
+class CommunityRecordInclusionNotificationBuilder(NotificationBuilder):
+
+    payload = [
+        RequestPayload,
+        CommunityPayload,
+        RecordPayload,
+    ]
+
+    recipients = [
+        RequestRecipients,
+        CommunityRecipients,
+        UserRecipient,
+    ]
+
+    backends = [
+        UserEmailBackend,
+    ]
+
+#
+# Payload
+#
+# invenio_requests/notifications.py
+class RequestPayload:
+    def __call__(self, request=None, **kwargs):
```

Comment:
overwriting call in this case would mean having things like `RequestPayload()()` which is not very intuitive. Is there a reason to have `__call__` overwriten instead of having a method return this values? I kind of like the `__call__` but also makes it not clear in interface for others to create a Payload class (is there an interface?).

Reply:
The pseudo-implementation for these has been updated (with some changed semantics as well of when they run). There's a more generic set of classes now, receiving a path/key for an entity reference to resolve. So the final usage looks something like:

```python
# usage in the NotificationBuilder class:
context = [
    EntityResolve(key="request"),
    EntityResolve(key="request.created_by"),
    EntityResolve(key="request.topic"),
    EntityResolve(key="request.receiver"),
]   

class EntityResolve:
    def __init__(self, key):
        self.key = key

    def __call__(self, notification: Notification):
        entity_ref = dict_lookup(notification, self.key)
        entity = EntityResolverRegistry.resolve_entity(entity_ref)
        dict_set(notification, self.key, entity)
        return notification
```

The idea is also to design the interface around simple `callable`s instead of some arbitrary `.{run,process,resolve}(...)` method.